File: ios/AniTracker/Player/PlayerModels.swift
import Foundation

/** 原生播放器当前所处的稳定阶段。 */
enum PlayerStatus: Equatable {
    case idle
    case loading
    case ready
    case playing
    case paused
    case buffering
    case ended
    case error
}

/** iOS 播放器支持的离散画面比例。 */
enum PlayerAspectRatio: String, CaseIterable, Identifiable {
    case automatic
    case widescreen
    case standard

    var id: String { rawValue }

    var title: String {
        switch self {
        case .automatic: "默认"
        case .widescreen: "16:9"
        case .standard: "4:3"
        }
    }

    var vlcValue: String? {
        switch self {
        case .automatic: nil
        case .widescreen: "16:9"
        case .standard: "4:3"
        }
    }
}

/** 一条可由 MobileVLCKit 加载的外部字幕。 */
struct PlayerSubtitle: Identifiable, Hashable {
    let id: String
    let label: String
    let url: URL
    let language: String?
    let isDefault: Bool
}

/** 移动播放列表中的单集媒体。 */
struct PlayerEpisode: Identifiable, Hashable {
    let id: String
    let title: String
    let episodeLabel: String
    let mediaURL: URL
    let durationMilliseconds: Int64
    let subtitles: [PlayerSubtitle]
}

/** 业务页面打开原生播放器时传递的完整参数。 */
struct PlayerLaunchRequest: Equatable {
    let sessionID: String
    let animeTitle: String
    let synopsis: String
    let artworkURL: URL?
    let episodes: [PlayerEpisode]
    let activeIndex: Int
    let startPositionMilliseconds: Int64
    let autoplay: Bool
}

/** MobileVLCKit 暴露给 SwiftUI 的音轨或字幕轨。 */
struct PlayerTrack: Identifiable, Equatable {
    let id: Int32
    let label: String
    let selected: Bool
}

/** SwiftUI 只依赖该快照，不直接读取 VLCMediaPlayer。 */
struct PlayerSnapshot: Equatable {
    var sessionID = ""
    var animeTitle = "Ani Tracker"
    var synopsis = ""
    var artworkURL: URL?
    var episodes: [PlayerEpisode] = []
    var activeIndex = 0
    var status: PlayerStatus = .idle
    var positionMilliseconds: Int64 = 0
    var durationMilliseconds: Int64 = 0
    var volume: Int32 = 70
    var muted = false
    var playbackRate: Float = 1
    var audioTracks: [PlayerTrack] = []
    var subtitleTracks: [PlayerTrack] = []
    var aspectRatio: PlayerAspectRatio = .automatic
    var watchedEpisodeIDs: Set<String> = []
    var autoNextSecondsRemaining: Int?
    var errorMessage: String?

    var activeEpisode: PlayerEpisode? {
        episodes.indices.contains(activeIndex) ? episodes[activeIndex] : nil
    }
}

File: android/app/src/main/java/dev/ani/tracker/android/player/PlayerModels.kt
package dev.ani.tracker.android.player

/** VLC 会话当前所处的稳定播放阶段。 */
enum class PlayerStatus {
    IDLE,
    LOADING,
    READY,
    PLAYING,
    PAUSED,
    BUFFERING,
    ENDED,
    ERROR
}

/** 内置播放器可加载的一条外部字幕。 */
data class PlayerSubtitle(
    val id: String,
    val label: String,
    val uri: String,
    val language: String? = null,
    val isDefault: Boolean = false
)

/** 移动端播放列表中的单集媒体。 */
data class PlayerEpisode(
    val id: String,
    val title: String,
    val episodeLabel: String,
    val uri: String,
    val durationMillis: Long = 0L,
    val subtitles: List<PlayerSubtitle> = emptyList()
)

/** 业务层交给原生播放器的完整启动参数。 */
data class PlayerLaunchRequest(
    val sessionId: String,
    val animeTitle: String,
    val description: String,
    val artworkUri: String?,
    val episodes: List<PlayerEpisode>,
    val activeIndex: Int,
    val startPositionMillis: Long,
    val autoplay: Boolean
)

/** VLC 暴露给 UI 的音轨或字幕轨。 */
data class PlayerTrack(
    val id: Int,
    val label: String,
    val selected: Boolean
)

/** 移动端播放器的单一状态源。 */
data class PlayerUiState(
    val sessionId: String = "",
    val animeTitle: String = "Ani Tracker",
    val description: String = "",
    val artworkUri: String? = null,
    val episodes: List<PlayerEpisode> = emptyList(),
    val activeIndex: Int = 0,
    val status: PlayerStatus = PlayerStatus.IDLE,
    val positionMillis: Long = 0L,
    val durationMillis: Long = 0L,
    val bufferPercent: Float = 0f,
    val volume: Int = 70,
    val muted: Boolean = false,
    val playbackRate: Float = 1f,
    val audioTracks: List<PlayerTrack> = emptyList(),
    val subtitleTracks: List<PlayerTrack> = emptyList(),
    val aspectRatio: String? = null,
    val errorMessage: String? = null
) {
    val activeEpisode: PlayerEpisode?
        get() = episodes.getOrNull(activeIndex)
}
